CHE-15 - Java stacktrace support (From Platform to Che Workspace)

vrubezhny opened the pull request:
Adds the possibility to click on a stacktrace line in order to open
a specified class in editor.

Example 1:
![Example 1](https://user-images.githubusercontent.com/620781/27247171-a75a03fa-52f6-11e7-9e11-5b4996b27214.gif)

Example 2:
![Example-2](https://user-images.githubusercontent.com/620781/27247320-bb3b0954-52f7-11e7-8b41-93c76d4313ba.gif)

Also fixes CHE-293 (#5001) - String Written to Dev-Machine stdout not Escaped Properly

Issue: https://issues.jboss.org/browse/CHE-15
Issue: https://issues.jboss.org/browse/CHE-293

see: https://github.com/eclipse/che/issues/5001

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests
-->

### What does this PR do?
Adds the possibility to click on a stacktrace line in order to open
a specified class in editor.
Also fixes CHE-293 (#5001) - String Written to Dev-Machine stdout not Escaped Properly

### What issues does this PR fix or reference?
Issue: https://issues.jboss.org/browse/CHE-15
Issue: https://issues.jboss.org/browse/CHE-293
(Issue: https://github.com/eclipse/che/issues/5001)


#### Changelog
<!-- one line entry to be added to changelog -->
The fix adds the possibility to click on a stacktrace line withing a console in order to open an according java class in editor and reveal it to the line according to the info that in provided in stacktrace line.

#### Release Notes
This version introduces a smart selection for the console outputs. When having exceptions or errors in your stacktrace, you can now directly jump into the file by clicking on it from the outputs. The editor will open with the file and point to the line corresponding to the one provided in the stacktrace. This allows to quickly navigate to the code which is raising an issue. 

[Animated Gif]

The smart selection capability is available for Java and NodeJS stacktrace at this moment. 

#### Docs PR
<!-- Please add a matching PR to [the docs repo](https://github.com/eclipse/che-docs) and link that PR to this issue.
Both will be merged at the same time. -->
Will need to be cover by: https://github.com/eclipse/che-docs/issues/249

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputConsole.java:
```diff
@@ -39,23 +41,30 @@
     /** Follow output when printing text */
     private boolean followOutput = true;
 
+    private OutputCustomizer					customizer = null;
```
hello, is that the indent is expected ?

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputConsole.java:
```diff
@@ -39,23 +41,30 @@
         view.hidePreview();
         view.setReRunButtonVisible(false);
         view.setStopButtonVisible(false);
+        
```
this line shouldn't be introduced

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputConsole.java:
```diff
@@ -39,23 +41,30 @@
+        
     }
-
+    
```
this line(spaces) shouldn't be introduced

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputConsole.java:
```diff
@@ -89,7 +98,7 @@ public void printText(String text) {
         view.print(text, text.endsWith("\r"), color);
 
-        for (ActionDelegate actionDelegate : actionDelegates) {
+    	for (ActionDelegate actionDelegate : actionDelegates) {
```
indent is with spaces, not with tabs

benoitf replied:
you've to check in the whole file as I see mainly only tabs while spaces are expected.

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,230 @@
+	private static RegExp LINE_AT_EXCEPTION = compile("(\\s+at address:.+)");
+	
+    private AppContext appContext;
+    private EditorAgent editorAgent;
```
the indent looks odd

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,230 @@
+import com.google.gwt.user.client.Timer;
+import com.google.inject.Inject;
+
+public class DefaultOutputCustomizer implements OutputCustomizer {
```
it is missing javadoc on the class. What it does, etc

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,230 @@
+public class DefaultOutputCustomizer implements OutputCustomizer {
+
+	private static RegExp LINE_AT = compile("(\\s+at .+)");
+	private static RegExp LINE_AT_EXCEPTION = compile("(\\s+at address:.+)");
```
it is missing the final keyword for the constants.

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,230 @@
+	 */
+	@Override
+	public String customize(String text) {
+    	String customText = text;
```
the formatting looks like it is not the expected one

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,230 @@
+    	String customText = text;
+    	
+    	MatchResult matcher = LINE_AT.exec(text);
+    	if(matcher != null) {
```
spaces around parenthesis

benoitf commented:
ci-build

benoitf commented:
as this issue is not a fix, it should include changelog and the release notes section.
@slemeur maybe documentation as well ?

vrubezhny commented:
@benoitf  I've added a couple of words to Change log. However, I don't know what is to be placed for Release Notes (especially once it not released yet)

benoitf commented:
@vrubezhny Changelog should be a one-line entry, so maybe move your "changelog" text as "Release notes" text and it should be ok

benoitf commented:
@vrubezhny also there was a compilation issue https://ci.codenvycorp.com/job/che-pullrequests-build/2833/

slemeur commented:
Hi @vrubezhny ! 
This seems really cool. 

@benoitf is right, this PR introduces a new feature so we need the release note section and the documentation prior to the merge. I can help on that. 
We'll also need the QA team to review and validate the changes: @musienko-maxim. 

Would you mind recording an animated gif to demo the new capability and attach it to the description of the PR?

vrubezhny commented:
@benoitf  The compilation is known: my to classes (intrf+impl) have "RedHat, Inc." in their headers instead of "Codenvy, S.A.". I don't feel like adding the whole che-core-ide-app as an exclusion to license-maven-plugin is a good approach. Is there a better solution? (Or I just have to use original CodeEnvy's header?)

vrubezhny commented:
@slemeur I've copied text of ChangeLog to Release Notes. Please let me know if it's not enough.

And I'll try to record an animated gif... (not sure if I can do it on Fedora, though)

vrubezhny commented:
@benoitf I believe I fixed formatting and other issues you pointed to, as well as the 'missing headers' one. Thanks for the review.

vrubezhny commented:
ci-build

benoitf commented:
ci-build

benoitf commented:
@vrubezhny as there is the use of regexp and parsing, IMHO a unit test should be added to cover the analysis of the stack trace.
Else it will be hard to track if it's still working or not upon refactoring.

vrubezhny commented:
@slemeur I've added the animated gifs to the PR description.

@benoitf The commit is updated due to the issue in my code, so ci-build is required for the PR once again. Sorry about that.

benoitf commented:
@vrubezhny Hello Vector. I'm not seing unit test as said yesterday. Also, could you add changes as adding a new commit (like using --fixup option) because, it's impossible to track what you changed between each review as we see only one commit even if you've done changes. Squashing/Rebasing can be done later when merging it. thanks.

vrubezhny commented:
@benoitf Yep, I haven't made JUnit test yet - I'll work on it on Monday. Thanks for pointing me to '--fixup' option, I'll follow this policy.

slemeur commented:
Hi @vrubezhny. Thanks for adding the animated gif, they are perfect. Looking at the code, I see that the regexp will properly cover java stacktraces. AFAIK it should also work for NodeJS but I'm bit worry that this might also create wrong interpretation for other type of stacktraces and bother the user. 

So to start the discussion, should we:
- check that the text found by the parser is actually a file before colorizing it and putting a link behind?
- should we also handle urls?
- allow the user to disable the smart selection in the console output
- have the ability for the user to add regular expressions (in preferences for example)?
- should we have the same behaviour in the terminal?

vrubezhny commented:
> Hi @vrubezhny. Thanks for adding the animated gif, they are perfect. Looking at the code, I see that the regexp will properly cover java stacktraces. AFAIK it should also work for NodeJS but I'm bit worry that this might also create wrong interpretation for other type of stacktraces and bother the user.

That's it. It's proposed to work with java-stacktraces printed out to a console, it's not intended to work with other kinds of stacktraces.  With proposed OutputCustomizer interface we can develop different kinds of customizer. Also, we can add preferences to allow enabling/disabling to the customizers, or configure different kinds of consoles to use different customizers.

> So to start the discussion, should we:
> - check that the text found by the parser is actually a file before colorizing it and putting a link behind?

We can't do that, first of all because of the performance reasons: it takes quite a lot of time (and I suppose a several asynchronous IO requests) to perform such a check. If we make it synchronously the console output would tend to freeze. Otherwise the text would not be printed in FIFO manner. As far as I know, desktop eclipse doesn't check for a file existence when a stacktrace is being printed out to a console, it performs searching when a user clicks on a colorized line.
The only thing I probably would verify (beside of matching the stacktrace line RegExp) is that file name really ends with ".java" to make sure that we're in "proper" stacktrace line.

> - should we also handle urls?

Not sure I understand right... Have an example of URLs appearance?

> - allow the user to disable the smart selection in the console output

Do you mean like "to disable" the console text customization (turning it to a link) and as such, not to provide a user with the possibility to search for a class by a stacktrace line? Not sure. I didn't see such a preference before.

> - have the ability for the user to add regular expressions (in preferences for example)?

If you're going to process the other kinds of stacktraces (like the javascript ones) then you should have the other kind of detection and post-processing (like parsing the line for packages/classes/file name/line number and/or adding, probably, javascript-to-java mappings to your search algorithm). This is not something that easy to realize as something that can be configured by preference and, imho, we should rather provide different customizers for users and let them a possibility to enable/disable them. Or hardly set up different customizers depending on the kind of console it should work on.  

> - should we have the same behaviour in the terminal?

As far as I understand, my change configures the terminal to use this customizer. So, if a java stacktrace would appear in terminal it will be customized as well. If it's not required I can try to disable/remove this behaviour from terminal.

vrubezhny commented:
@benoitf I'm going to add a commit with JUnit test to this PR. Should I use that `--fixup` option, or a separate commit would be OK?

benoitf commented:
@vrubezhny it is up to you, in all cases it will add a new commit (fixup is generating a commit as well)
At the end of the review you can either automatically rebase if you used fixup or manually rebase 
Or we squash commits when merging on github

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,234 @@
+                customText += text.substring(openBracket + "(".length(), closingBracket);
+                customText += "</a>";
+                customText += text.substring(closingBracket);
+                text = customText;
```
do you think that a String.format would be easier to read than concatenating tons of stuff ?

benoitf replied:
so bad idea as GWT is not supporting formatting 👎

vrubezhny replied:
String.format/MessageFormat.format/Formatter.format are not available for GWT application and it doesn't look like a GWT replacement exists.

benoitf commented on ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/console/DefaultOutputCustomizer.java:
```diff
@@ -0,0 +1,234 @@
+            public List<File> apply(Resource[] children) throws FunctionException {
+                final List<File> result = newArrayList();
+
+                for (Resource child : children) {
```
would it be simpler to read with streams / filter/ Collectors.toList() ?

benoitf commented on ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/console/DefaultOutputCustomizerTest.java:
```diff
@@ -0,0 +1,100 @@
+    public void testNonStackTraceLines() throws Exception {
+        testStackTraceLine(false, "[STDOUT] Listening for transport dt_socket at address: 4403", null);
+    }
+}
```
an empty blank line is required at the end of the file

vrubezhny commented:
@benoitf I've added fixups for:
- turning org.eclipse.che.ide.console.DefaultOutputCustomizer.collectChildren(Container, Path) to use streams / filter/ Collectors.toList() 
- added new line to last line of DefaultOutputCustomizerTest
- added new line before DefaultOutputCustomizer.collectChildren() method's closing bracket (somehow lost on formatting)

benoitf commented:
@vparfonov ?
